velox/functions/lib/QuantileDigest.h

```diff
@@ -0,0 +1,850 @@
+              1);
+  } else {
+    return root_ == -1
+        ? 1
+        : std::max(
+              static_cast<int>(
+                  static_cast<double>(
+                      (levels_[root_] == 32 ? 64 : levels_[root_]) + 1) /
+                  maxError_),
+              1);
+  }
+}
+
+template <typename T>
+void QuantileDigest<T>::insert(U value, double count) {
+  if (count < qdigest::kZeroWeightThreshold) {
+    return;
+  }
+  U lastBranch = 0;
+  int32_t parent = -1;
+  int32_t current = root_;
+  while (current != -1) {
+    auto currentValue = values_[current];
+    auto currentLevel = levels_[current];
+    if (!inSameSubtree(value, currentValue, currentLevel)) {
+      setChild(
+          parent, lastBranch, makeSiblings(current, createLeaf(value, count)));
+      return;
+    }
+
+    if (currentLevel == 0 && currentValue == value) {
+      counts_[current] += count;
+      weightedCount_ += count;
+      return;
+    }
+
+    U branch = value & getBranchMask(currentLevel);
+    parent = current;
+    lastBranch = branch;
+    if (branch == 0) {
+      current = lefts_[current];
+    } else {
+      current = rights_[current];
+    }
+  }
+  setChild(parent, lastBranch, createLeaf(value, count));
+}
+
+template <typename T>
+int32_t QuantileDigest<T>::createLeaf(U value, double count) {
+  return createNode(value, 0, count);
+}
+
+template <typename T>
+int32_t QuantileDigest<T>::createNode(U value, int8_t level, double count) {
+  auto node = popFree();
+  if (node == -1) {
+    if (nextNode_ == counts_.size()) {
+      int32_t newSize = counts_.size() +
+          std::min(counts_.size(),
+                   static_cast<uint64_t>(calculateCompressionFactor() / 5 + 1));
+      counts_.resize(newSize);
+      levels_.resize(newSize);
+      values_.resize(newSize);
+      lefts_.resize(newSize);
+      rights_.resize(newSize);
+    }
+
+    node = nextNode_;
+    nextNode_++;
+  }
+  weightedCount_ += count;
+  values_[node] = value;
+  levels_[node] = level;
+  counts_[node] = count;
+  lefts_[node] = -1;
+  rights_[node] = -1;
+  return node;
+}
+
+template <typename T>
+bool QuantileDigest<T>::inSameSubtree(U bitsA, U bitsB, int32_t level) {
+  if constexpr (std::is_same_v<U, int64_t>) {
+    return (level == 64) ||
+        ((static_cast<uint64_t>(bitsA) >> level) ==
+         (static_cast<uint64_t>(bitsB) >> level));
+  } else {
+    return (level == 32) ||
+        ((static_cast<uint32_t>(bitsA) >> level) ==
+         (static_cast<uint32_t>(bitsB) >> level));
+  }
+}
+
+template <typename T>
+QuantileDigest<T>::U QuantileDigest<T>::getBranchMask(int8_t level) {
+  return static_cast<U>(1) << (level - 1);
+}
+
+template <typename T>
+int32_t QuantileDigest<T>::makeSiblings(int32_t first, int32_t second) {
+  auto firstValue = values_[first];
+  auto secondValue = values_[second];
+  int32_t parentLevel;
+  if constexpr (std::is_same_v<U, int64_t>) {
```

Comment:
This can be extracted to its own method.  Basically anything we need to dispatch on `U` extract that to an helper